### PR TITLE
feat(attribute_value): Don't target NAN, INF, -INF and  > 2^53

### PR DESCRIPTION
**Pull request by oakbani**
Summary
-------
Don't target NAN, INF or -INF doubles
Don't target Values > 2^53. 
Modified is_attribute_valid to check if the number is finite. 

Test plan
---------

Issues
------
OASIS-3654
https://github.com/optimizely/python-sdk/pull/146#discussion_r237850806

### CODE REVIEW

**Comment by oakbani** [comment]
Let me know if I should extend is_finite_check to revenue and numeric values within this PR. Looks like a PR on parsing revenue value is already pending for a while. https://github.com/optimizely/python-sdk/pull/134

File: optimizely/helpers/validator.py [hunk 1]
```diff
@@ -185,7 +187,40 @@ def is_attribute_valid(attribute_key, attribute_value):
...
+  if not isinstance(value, (numbers.Integral, float)):
+      # numbers.Integral instead of int to accomodate long integer in python 2
+    return False
+
+  if isinstance(value, bool):
+    # bool is a subclass of int
+    return False
+
+  if isinstance(value, numbers.Integral):
+    if value > 1e53:
```
Comment (nchilada) [inline optimizely/helpers/validator.py]:
```suggestion
    if abs(value) > 1e53:
```

File: optimizely/helpers/validator.py [hunk 2]
```diff
@@ -185,7 +187,40 @@ def is_attribute_valid(attribute_key, attribute_value):
...
+
+  if not isinstance(value, (numbers.Integral, float)):
+      # numbers.Integral instead of int to accomodate long integer in python 2
+    return False
+
+  if isinstance(value, bool):
+    # bool is a subclass of int
+    return False
+
+  if isinstance(value, numbers.Integral):
```
Comment (nchilada) [inline optimizely/helpers/validator.py]: We should apply the size limit even if the value is a float. Can we get rid of this gate?

File: optimizely/helpers/validator.py [hunk 3]
```diff
@@ -185,7 +187,40 @@ def is_attribute_valid(attribute_key, attribute_value):
...
+  if isinstance(attribute_value, (string_types, bool)):
     return True
 
+  if isinstance(attribute_value, (numbers.Integral, float)):
+    return is_finite_number(attribute_value)
+
   return False
+
+
+def is_finite_number(value):
```
Comment (nchilada) [inline optimizely/helpers/validator.py]: Are you planning to merge this PR on its own and then update #146 to call `is_finite_number` during audience evaluation? That works for me!

File: tests/helpers_tests/test_validator.py [hunk 4]
```diff
@@ -168,6 +171,55 @@ def test_is_attribute_valid(self):
...
+
+    with mock.patch('optimizely.helpers.validator.is_finite_number',
+                    return_value=False) as is_finite:
+      self.assertFalse(validator.is_attribute_valid('test_attribute', 5.5))
+
+    is_finite.assert_called_once_with(5.5)
+
+    with mock.patch('optimizely.helpers.validator.is_finite_number',
+                    return_value='abc') as is_finite:
+      self.assertEqual('abc', validator.is_attribute_valid('test_attribute', 0))
```
Comment (nchilada) [inline tests/helpers_tests/test_validator.py]: Nit: I'm not sure this block is needed.

File: tests/helpers_tests/test_validator.py [hunk 5]
```diff
@@ -168,6 +171,55 @@ def test_is_attribute_valid(self):
...
+    self.assertFalse(validator.is_finite_number([]))
+    self.assertFalse(validator.is_finite_number(()))
+
+    # test invalid numbers
+    self.assertFalse(validator.is_finite_number((float('inf'))))
+    self.assertFalse(validator.is_finite_number((float('-inf'))))
+    self.assertFalse(validator.is_finite_number((float('nan'))))
+    self.assertFalse(validator.is_finite_number(int(1e53) + 1))
+    if PY2:
+      self.assertFalse(validator.is_finite_number(long(1e53) + 1))
```
Comment (nchilada) [inline tests/helpers_tests/test_validator.py]: As implied by my comments in the code under test, we also want assertions like these:
```python
self.assertFalse(validator.is_finite_number(float(1e53) + 1.0))
self.assertFalse(validator.is_finite_number(-int(1e53) - 1))
if PY2:
  self.assertFalse(validator.is_finite_number(-long(1e53) - 1))
self.assertFalse(validator.is_finite_number(-float(1e53) - 1.0))
```
(maybe reordered as you deem fit)

File: tests/helpers_tests/test_validator.py [hunk 6]
```diff
@@ -168,6 +171,55 @@ def test_is_attribute_valid(self):
...
+
+    # test valid numbers
+    self.assertTrue(validator.is_finite_number((0)))
+    self.assertTrue(validator.is_finite_number((5)))
+    self.assertTrue(validator.is_finite_number((5.5)))
+    self.assertTrue(validator.is_finite_number((float(1e53) + 1)))
+    self.assertTrue(validator.is_finite_number((int(1e53))))
+    if PY2:
+      self.assertTrue(validator.is_finite_number((long(1e53))))
```
Comment (nchilada) [inline tests/helpers_tests/test_validator.py]: And
```python
self.assertTrue(validator.is_finite_number((float(1e53))))
```

File: tests/helpers_tests/test_validator.py [hunk 7]
```diff
@@ -168,6 +171,55 @@ def test_is_attribute_valid(self):
...
+    self.assertFalse(validator.is_finite_number('HelloWorld'))
+    self.assertFalse(validator.is_finite_number(True))
+    self.assertFalse(validator.is_finite_number(False))
+    self.assertFalse(validator.is_finite_number(None))
+    self.assertFalse(validator.is_finite_number({}))
+    self.assertFalse(validator.is_finite_number([]))
+    self.assertFalse(validator.is_finite_number(()))
+
+    # test invalid numbers
+    self.assertFalse(validator.is_finite_number((float('inf'))))
```
Comment (nchilada) [inline tests/helpers_tests/test_validator.py]: Nit: there are unnecessary double parens in a lot of these assertions!
```suggestion
    self.assertFalse(validator.is_finite_number(float('inf')))
```

File: optimizely/helpers/validator.py (hunk 1 above)
Comment (nchilada) [reply optimizely/helpers/validator.py]: Ugh - I just realized that I asked for the limit to be `1e53`/`10^53`, but it should actually be `2^53`. Do you think you could make that change too?

File: optimizely/helpers/validator.py (hunk 3 above)
Comment (oakbani) [reply optimizely/helpers/validator.py]: Yeah. Once this is approved. I will merge it changes to #146

Comment (nchilada) [reply optimizely/helpers/validator.py]: Oops, looks like #146 is merged! But it actually added the calls to `is_finite_number`. So I guess you just need to rebase this branch onto devel in order to add the 2^53/NaN/Infinity limits.

**Comment by nchilada** [comment]
> > Let me know if I should extend is_finite_check to revenue and numeric values within this PR. Looks like a PR on parsing revenue value is already pending for a while. #134
> 
> We can only effectively accept integers for `revenue`, so I'm thinking we should probably just close #134. 😕
> 
> I'm asking offline to figure out the exact size limits for an event's `revenue`/`value`/tags. Sorry for not having an immediate answer!

Let's skip this after all; it's taking a while to get an answer, and it's not directly related to our current epic, and it looks like we're skipping this in other SDKs (e.g. optimizely/objective-c-sdk#348).

File: tests/helpers_tests/test_validator.py [hunk 8]
```diff
@@ -168,6 +171,61 @@ def test_is_attribute_valid(self):
...
+    self.assertFalse(validator.is_finite_number(-float(2**53) - 2.0))
+    if PY2:
+      self.assertFalse(validator.is_finite_number(long(2**53) + 1))
+      self.assertFalse(validator.is_finite_number(-long(2**53) - 1))
+
+    # test valid numbers
+    self.assertTrue(validator.is_finite_number(0))
+    self.assertTrue(validator.is_finite_number(5))
+    self.assertTrue(validator.is_finite_number(5.5))
+    self.assertTrue(validator.is_finite_number(float(2**53) + 1.0))
```
Comment (oakbani) [inline tests/helpers_tests/test_validator.py]: https://stackoverflow.com/questions/1848700/biggest-integer-that-can-be-stored-in-a-double

File: tests/helpers_tests/test_condition.py [hunk 9]
```diff
@@ -371,6 +429,42 @@ def test_greater_than_int__returns_true__when_user_value_greater_than_condition_
...
+    if PY2:
+      evaluator = condition_helper.CustomAttributeConditionEvaluator(
+        gt_int_condition_list, {'meters_travelled': long(49)}
+      )
+
+      with mock.patch('optimizely.helpers.validator.is_finite_number',
+                       side_effect=side_effect) as is_finite:
+        self.assertIsNone(evaluator.evaluate(0))
+      is_finite.assert_called_with(long(49))
+
```
Comment (nchilada) [inline tests/helpers_tests/test_condition.py]: Ooh, I like the idea of using `side_effect` here. Can we go farther with this and test that `is_finite` only needs to reject the condition value _or_ the user attribute value in order to stop condition evaluation?

Something like
```python
def test_greater_than__calls_is_finite_number(self):
  evaluator = condition_helper.CustomAttributeConditionEvaluator(
    gt_int_condition_list, {'meters_travelled': 48.1}
  )
  
  def is_finite_number__rejecting_condition_value(value):
    if value == 48:
      return False
    return True
  
  with mock.patch('optimizely.helpers.validator.is_finite_number',
                  side_effect=is_finite_number__rejecting_condition_value):
    self.assertIsNone(evaluator.evaluate(0))
  
  def is_finite_number__rejecting_user_attribute_value(value):
    if value == 48.1:
      return False
    return True
  
  with mock.patch('optimizely.helpers.validator.is_finite_number',
                  side_effect=is_finite_number__rejecting_user_attribute_value):
    self.assertIsNone(evaluator.evaluate(0))
  
  def is_finite_number__accepting_both_values(value):
    return True
  
  with mock.patch('optimizely.helpers.validator.is_finite_number',
                  side_effect=is_finite_number__accepting_both_values):
    self.assertIsNone(evaluator.evaluate(0))
```

I don't think we need to test ints, longs, and floats separately... not here, at least... we can probably get away with `test_greater_than__calls_is_finite_number`, `test_less_than__calls_is_finite_number`, and maybe ``test_exact__given_number_values__calls_is_finite_number`. If that makes sense.

@oakbani @aliabbasrizvi what do you think?

Comment (oakbani) [reply tests/helpers_tests/test_condition.py]: Nice suggestion. Revised accordingly.

File: optimizely/helpers/validator.py [hunk 10]
```diff
@@ -187,20 +187,24 @@ def is_attribute_valid(attribute_key, attribute_value):
...
-  """ Method to validate if the given value is a number and not one of NAN, INF, -INF.
+  """ Validates if the given value is a number, enforces
+   absolute limit of 2^53 and restricts NAN, INF, -INF.
 
   Args:
     value: Value to be validated.
 
   Returns:
-    Boolean: True if value is a number and not NAN, INF or -INF else False.
+    Boolean: True if value is a number and not NAN, INF, -INF or greater than 2^53 else False.
```
Comment (msohailhussain) [inline optimizely/helpers/validator.py]: nit: or absolute limit of 2^53

Comment (aliabbasrizvi) [reply optimizely/helpers/validator.py]: Late to the party, but wouldn't I want this limit for Double type variables and Integer (sys.maxint) for Integer variables?

Comment (aliabbasrizvi) [reply optimizely/helpers/validator.py]: @nchilada ^

Comment (aliabbasrizvi) [reply optimizely/helpers/validator.py]: I think I get it. I am mixing feature variables here. Apologies.

Comment (nchilada) [reply optimizely/helpers/validator.py]: @aliabbasrizvi it's true that Python integers can have larger values than 2^53, but they're passed to the backend as JSON Numbers which are best limited to 2^53. Does that clarify why we limit integers too?

File: tests/helpers_tests/test_condition.py [hunk 11]
```diff
@@ -290,19 +290,27 @@ def test_exact__given_number_values__calls_is_finite_number(self):
         exact_int_condition_list, {'lasers_count': 9000}
       )
 
+    # assert that isFiniteNumber only needs to reject condition value to stop evaluation.
     with mock.patch('optimizely.helpers.validator.is_finite_number',
-                    return_value=True) as mock_is_finite:
-      self.assertTrue(evaluator.evaluate(0))
-    mock_is_finite.assert_called_with(9000)
+                    side_effect=[False, True]) as mock_is_finite:
```
Comment (oakbani) [inline tests/helpers_tests/test_condition.py]: @nchilada In exact match need to send exactly same values to ensure that it's the mocked is_finite that is really driving the result. Therefore, can't make a value-base side-effect method as in gt/lt methods. 
This returns False on 1st call and True on 2nd.

Comment (nchilada) [reply tests/helpers_tests/test_condition.py]: Ah, you're right. Good solution. Thanks @oakbani!

**Pull request merged**